tables/core/array.py

```diff
@@ -42,20 +45,46 @@ def get_enum(self):
 
         return self.atom.enum
 
-    def read(self, start=None, stop=None, step=None):
-        (start, stop, step) = self._process_range_read(start, stop, step)
-        if self.nrows == 1:
-            return numpy.array(self[()])
+    def read(self, start=None, stop=None, step=None, out=None):
+
+        # Scalar dataset
+        if self.shape == ():
+            arr = self[()]
+            nrowstoread = 1
         else:
-            return self[start:stop:step]
+            (start, stop, step) = self._process_range_read(start, stop, step)
+            arr = self[start:stop:step]
+            nrowstoread = len(range(start, stop, step))
+
+        if (isinstance(arr, np.ndarray) and byteorders[arr.dtype.byteorder] != sys.byteorder):
+            arr = arr.byteswap(True)
+            arr.dtype = arr.dtype.newbyteorder()
+        if out is not None:
+            if self.flavor != 'numpy':
+                msg = ("Optional 'out' argument may only be supplied if array "
+                       "flavor is 'numpy', currently is {0}").format(self.flavor)
+                raise TypeError(msg)
+            bytes_required = self.rowsize * nrowstoread
+            # if buffer is too small, it will segfault
+            if bytes_required != out.nbytes:
+                raise ValueError(('output array size invalid, got {0} bytes, '
+                                  'need {1} bytes').format(out.nbytes,
+                                                           bytes_required))
+            if not out.flags['C_CONTIGUOUS']:
+                raise ValueError('output array not C contiguous')
+
+            np.copyto(out, arr)
```

Comment:
You can avoid the `arr` temporary by using [`read_direct(array, source_sel=None, dest_sel=None)`](http://docs.h5py.org/en/latest/high/dataset.html).